src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -305,43 +296,84 @@ func (c controller) unallocatedMachines(bmhList metal3v1alpha1.BareMetalHostList
+		// to indicated that the statusAnnotation is available. This is only on
+		// the master nodes.
+		if _, ok := annotations[metal3v1alpha1.PausedAnnotation]; ok {
+			c.log.Infof("Removing pausedAnnotation from BMH host %s", bmh.Name)
+			delete(annotations, metal3v1alpha1.PausedAnnotation)
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if annotations[metal3v1alpha1.StatusAnnotation] == "" {
+		c.log.Infof("Skipping setting status of BMH host %s, status annotation not present", bmh.Name)
+		return false, nil
+	}
+	objStatus, err := c.unmarshalStatusAnnotation(content)
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to unmarshal status annotation of %s", bmh.Name)
+		return false, err
+	}
+	bmh.Status = *objStatus
+	if bmh.Status.LastUpdated.IsZero() {
+		// Ensure the LastUpdated timestamp in set to avoid
+		// infinite loops if the annotation only contained
+		// part of the status information.
+		t := metav1.Now()
+		bmh.Status.LastUpdated = &t
+	}
+	err = c.kc.UpdateBMHStatus(bmh)
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to update status of BMH %s", bmh.Name)
+		return false, err
+	}
+	delete(annotations, metal3v1alpha1.StatusAnnotation)
+	return true, nil
+}
+
+func (c controller) updateBMHs(bmhList *metal3v1alpha1.BareMetalHostList, machineList *mapiv1beta1.MachineList) bool {
 	allUpdated := true
 
+	provisioningExists, err := c.kc.IsMetalProvisioningExists()
+	if err != nil {
+		return false
+	}
+
 	for i := range bmhList.Items {
 		bmh := bmhList.Items[i]
 		c.log.Infof("Checking bmh %s", bmh.Name)
-		annotations := bmh.GetAnnotations()
-		content := []byte(annotations[metal3v1alpha1.StatusAnnotation])
 
-		if annotations[metal3v1alpha1.StatusAnnotation] == "" && bmh.Spec.ConsumerRef != nil {
-			c.log.Infof("Skipping setting status of BMH host %s, status annotation not present", bmh.Name)
-			continue
-		}
-		allUpdated = false
-		if annotations[metal3v1alpha1.StatusAnnotation] != "" {
-			objStatus, err := c.unmarshalStatusAnnotation(content)
-			if err != nil {
-				c.log.WithError(err).Errorf("Failed to unmarshal status annotation of %s", bmh.Name)
-				continue
-			}
-			bmh.Status = *objStatus
-			if bmh.Status.LastUpdated.IsZero() {
-				// Ensure the LastUpdated timestamp in set to avoid
-				// infinite loops if the annotation only contained
-				// part of the status information.
-				t := metav1.Now()
-				bmh.Status.LastUpdated = &t
-			}
-			err = c.kc.UpdateBMHStatus(&bmh)
-			if err != nil {
-				c.log.WithError(err).Errorf("Failed to update status of BMH %s", bmh.Name)
+		needsUpdate := false
+		if provisioningExists {
+			if bmh.Status.Provisioning.State == metal3v1alpha1.StateNone || bmh.Status.Provisioning.State == metal3v1alpha1.StateUnmanaged {
+				needsUpdate = !bmh.Spec.ExternallyProvisioned
+				bmh.Spec.ExternallyProvisioned = true
+			} else {
+				c.log.Infof("bmh %s Provisioning.State=%s", bmh.Name, bmh.Status.Provisioning.State)
+				allUpdated = false
```

Comment:
I don't think you want to return false in this case - if the Host isn't in one of those states it's never going to be (the state machine doesn't move back in that direction), so we'll be stuck in an infinite loop.